opensearch/src/main/java/org/opensearch/sql/opensearch/request/PredicateAnalyzer.java

```diff
@@ -1190,6 +1259,171 @@ private static String timestampValueForPushDown(String value) {
     // https://github.com/opensearch-project/sql/pull/3442
   }
 
+  public static class ScriptQueryExpression extends QueryExpression {
+    private final String code;
+
+    public ScriptQueryExpression(
+        RexNode rexNode,
+        RelDataType rowType,
+        Map<String, ExprType> fieldTypes,
+        RelOptCluster cluster) {
+      ReferenceFieldVisitor validator = new ReferenceFieldVisitor(rowType, fieldTypes, true);
+      // Dry run visitInputRef to make sure the input reference ExprType is valid for script
+      // pushdown
+      validator.visitEach(List.of(rexNode));
+      RelJsonSerializer serializer = new RelJsonSerializer(cluster);
+      this.code = serializer.serialize(rexNode, rowType, fieldTypes);
```

Comment:
ScriptQueryExpression() is invoked multiple times during the optimization stage. Can we defer code generation until the implementation stage?